engine/src/main/java/org/hibernate/search/backend/impl/InTransactionWorkQueueSynchronization.java

```diff
@@ -0,0 +1,133 @@
+/*
+ * Hibernate Search, full-text search for your domain model
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+
+package org.hibernate.search.backend.impl;
+
+import java.util.concurrent.ConcurrentMap;
+import javax.transaction.Status;
+
+import org.hibernate.search.backend.spi.Work;
+import org.hibernate.search.engine.integration.impl.ExtendedSearchIntegrator;
+import org.hibernate.search.util.logging.impl.Log;
+import org.hibernate.search.util.logging.impl.LoggerFactory;
+
+/**
+ * Execute final work inside a transaction.
+ *
+ * @author Emmanuel Bernard
+ */
+public class InTransactionWorkQueueSynchronization implements WorkQueueSynchronization {
+
+	private static final Log log = LoggerFactory.make();
+
+	/**
+	 * FullTextIndexEventListener is using a WeakIdentityHashMap<Session,Synchronization>
+	 * So make sure all Synchronization implementations don't have any
+	 * (direct or indirect) reference to the Session.
+	 */
+
+	private final QueueingProcessor queueingProcessor;
+	private boolean consumed;
+	private boolean prepared;
+	private final ConcurrentMap<Object, WorkQueueSynchronization> queuePerTransaction;
+	private final WorkQueue queue;
+	private final Object transactionIdentifier;
+
+	/**
+	 * in transaction work
+	 */
+	public InTransactionWorkQueueSynchronization(Object transactionIdentifier, QueueingProcessor queueingProcessor,
+												ConcurrentMap<Object, WorkQueueSynchronization> queuePerTransaction,
+												ExtendedSearchIntegrator extendedIntegrator) {
+		this.transactionIdentifier = transactionIdentifier;
+		this.queueingProcessor = queueingProcessor;
+		this.queuePerTransaction = queuePerTransaction;
+		queue = new WorkQueue( extendedIntegrator );
+	}
+
+	public void add(Work work) {
+		queueingProcessor.add( work, queue );
+	}
+
+	public boolean isConsumed() {
+		return consumed;
+	}
+
+	@Override
+	public void beforeCompletion() {
+		// we are doing all the work in the before completion phase so that it is part of the transaction
+		try {
+			if ( prepared ) {
+				if ( log.isTraceEnabled() ) {
```

Comment:
I've added a commit to do this @gunnarmorling. Let me know if that's not what you had in mind.